File: packages/core/src/PromptRenderer.ts
/**
 * @module minions-prompts/PromptRenderer
 * Variable interpolation engine for prompt templates.
 * Supports {{variable}}, {{#if condition}}, and {{#each items}} syntax.
 */

/** Error thrown when rendering fails due to missing or invalid variables. */
export class RendererError extends Error {
  constructor(
    message: string,
    public readonly missingVariables?: string[],
  ) {
    super(message);
    this.name = 'RendererError';
  }
}

/** Options for the render method. */
export interface RenderOptions {
  /** If true, missing variables throw a RendererError. Default: true. */
  strict?: boolean;
  /** Variables declared as required; missing ones always throw regardless of strict mode. */
  requiredVariables?: string[];
}

/**
 * Renders prompt templates with variable substitution and block support.
 *
 * @example
 * ```typescript
 * const renderer = new PromptRenderer();
 * const result = renderer.render(
 *   'Hello, {{name}}! You are a {{role}}.',
 *   { name: 'Alice', role: 'developer' }
 * );
 * // → 'Hello, Alice! You are a developer.'
 * ```
 */
export class PromptRenderer {
  /**
   * Renders a prompt template with the given variables.
   *
   * Supports:
   * - `{{variable}}` — basic substitution
   * - `{{#if variable}}...{{/if}}` — conditional blocks
   * - `{{#each array}}...{{/each}}` — iteration blocks
   *
   * @param template - The prompt template string.
   * @param variables - Key/value pairs to substitute.
   * @param options - Rendering options.
   * @returns The rendered string.
   * @throws {RendererError} If required variables are missing.
   */
  render(
    template: string,
    variables: Record<string, unknown> = {},
    options: RenderOptions = {},
  ): string {
    const { strict = true, requiredVariables = [] } = options;

    // Check required variables
    const missing = requiredVariables.filter(
      (v) => variables[v] === undefined || variables[v] === null,
    );
    if (missing.length > 0) {
      throw new RendererError(
        `Missing required variables: ${missing.join(', ')}`,
        missing,
      );
    }

    let result = template;

    // Process #each blocks first
    result = this.processEachBlocks(result, variables, strict);

    // Process #if blocks
    result = this.processIfBlocks(result, variables, strict);

    // Process simple variable substitutions
    result = this.processVariables(result, variables, strict);

    // Unescape literal \{{ sequences
    result = result.replace(/\\{{/g, '{{');

    return result;
  }

  /**
   * Extracts all variable names referenced in a template.
   *
   * @param template - The prompt template string.
   * @returns Array of unique variable names.
   */
  extractVariables(template: string): string[] {
    const names = new Set<string>();

    // Match {{variable}} patterns (not # or / prefixed)
    const varPattern = /\{\{([^#/][^}]*?)\}\}/g;
    let match: RegExpExecArray | null;
    while ((match = varPattern.exec(template)) !== null) {
      names.add(match[1]!.trim());
    }

    // Match condition names in #if blocks
    const ifPattern = /\{\{#if\s+([^}]+)\}\}/g;
    while ((match = ifPattern.exec(template)) !== null) {
      names.add(match[1]!.trim());
    }

    // Match collection names in #each blocks
    const eachPattern = /\{\{#each\s+([^}]+)\}\}/g;
    while ((match = eachPattern.exec(template)) !== null) {
      names.add(match[1]!.trim());
    }

    return Array.from(names);
  }

  private processEachBlocks(
    template: string,
    variables: Record<string, unknown>,
    strict: boolean,
  ): string {
    const eachPattern = /\{\{#each\s+(\w+)\}\}([\s\S]*?)\{\{\/each\}\}/g;
    return template.replace(eachPattern, (_, name: string, body: string) => {
      const collection = variables[name];
      if (collection === undefined || collection === null) {
        if (strict) {
          // Not required, just return empty
        }
        return '';
      }
      if (!Array.isArray(collection)) {
        return '';
      }
      return collection
        .map((item) => {
          if (typeof item === 'object' && item !== null) {
            // Object iteration: expose item fields as variables
            return this.processVariables(body, item as Record<string, unknown>, false);
          }
          // Primitive iteration: expose as {{this}}
          return body.replace(/\{\{this\}\}/g, String(item));
        })
        .join('');
    });
  }

  private processIfBlocks(
    template: string,
    variables: Record<string, unknown>,
    _strict: boolean,
  ): string {
    const ifPattern = /\{\{#if\s+(\w+)\}\}([\s\S]*?)\{\{\/if\}\}/g;
    return template.replace(ifPattern, (_, name: string, body: string) => {
      const value = variables[name];
      if (this.isTruthy(value)) {
        return body;
      }
      return '';
    });
  }

  private processVariables(
    template: string,
    variables: Record<string, unknown>,
    strict: boolean,
  ): string {
    return template.replace(/(?<!\\)\{\{([^#/][^}]*?)\}\}/g, (match, name: string) => {
      const key = name.trim();
      if (key in variables) {
        const val = variables[key];
        if (val === null || val === undefined) return '';
        if (typeof val === 'object') return JSON.stringify(val);
        return String(val);
      }
      if (strict) {
        // Leave unknown vars as-is in non-strict mode
        return match;
      }
      return match;
    });
  }

  private isTruthy(value: unknown): boolean {
    if (value === null || value === undefined) return false;
    if (value === false) return false;
    if (value === '') return false;
    if (Array.isArray(value) && value.length === 0) return false;
    return true;
  }
}

File: packages/python/minions_prompts/prompt_renderer.py
"""
PromptRenderer — variable interpolation engine for prompt templates.
"""

from __future__ import annotations

import re
import json
from typing import Any


class RendererError(Exception):
    """Raised when rendering fails due to missing or invalid variables.

    Attributes:
        missing_variables: List of variable names that were missing.
    """

    def __init__(self, message: str, missing_variables: list[str] | None = None) -> None:
        super().__init__(message)
        self.missing_variables = missing_variables or []


class PromptRenderer:
    """Renders prompt templates with variable substitution and block support.

    Supports:
    - ``{{variable}}`` — basic substitution
    - ``{{#if variable}}...{{/if}}`` — conditional blocks
    - ``{{#each array}}...{{/each}}`` — iteration blocks

    Example::

        renderer = PromptRenderer()
        result = renderer.render(
            "Hello, {{name}}! You are a {{role}}.",
            {"name": "Alice", "role": "developer"},
        )
        # → "Hello, Alice! You are a developer."
    """

    def render(
        self,
        template: str,
        variables: dict[str, Any] | None = None,
        *,
        strict: bool = True,
        required_variables: list[str] | None = None,
    ) -> str:
        """Render a prompt template with the given variables.

        Args:
            template: The prompt template string.
            variables: Key/value pairs to substitute.
            strict: If True (default), leave unknown placeholders as-is.
            required_variables: Variables that must be present; raises RendererError if missing.

        Returns:
            The rendered string.

        Raises:
            RendererError: If required variables are missing.
        """
        variables = variables or {}
        required_variables = required_variables or []

        missing = [v for v in required_variables if variables.get(v) is None]
        if missing:
            raise RendererError(
                f"Missing required variables: {', '.join(missing)}",
                missing_variables=missing,
            )

        result = template
        result = self._process_each_blocks(result, variables)
        result = self._process_if_blocks(result, variables)
        result = self._process_variables(result, variables)
        result = result.replace(r"\{{", "{{")
        return result

    def extract_variables(self, template: str) -> list[str]:
        """Extract all variable names referenced in a template.

        Args:
            template: The prompt template string.

        Returns:
            List of unique variable names.
        """
        names: set[str] = set()

        for m in re.finditer(r"\{\{([^#/][^}]*?)\}\}", template):
            names.add(m.group(1).strip())

        for m in re.finditer(r"\{\{#if\s+(\w+)\}\}", template):
            names.add(m.group(1).strip())

        for m in re.finditer(r"\{\{#each\s+(\w+)\}\}", template):
            names.add(m.group(1).strip())

        return list(names)

    def _process_each_blocks(self, template: str, variables: dict[str, Any]) -> str:
        def replace_each(match: re.Match) -> str:
            name = match.group(1)
            body = match.group(2)
            collection = variables.get(name)
            if not isinstance(collection, list):
                return ""
            parts = []
            for item in collection:
                if isinstance(item, dict):
                    parts.append(self._process_variables(body, item))
                else:
                    parts.append(body.replace("{{this}}", str(item)))
            return "".join(parts)

        return re.sub(r"\{\{#each\s+(\w+)\}\}([\s\S]*?)\{\{/each\}\}", replace_each, template)

    def _process_if_blocks(self, template: str, variables: dict[str, Any]) -> str:
        def replace_if(match: re.Match) -> str:
            name = match.group(1)
            body = match.group(2)
            value = variables.get(name)
            return body if self._is_truthy(value) else ""

        return re.sub(r"\{\{#if\s+(\w+)\}\}([\s\S]*?)\{\{/if\}\}", replace_if, template)

    def _process_variables(self, template: str, variables: dict[str, Any]) -> str:
        def replace_var(match: re.Match) -> str:
            key = match.group(1).strip()
            if key in variables:
                val = variables[key]
                if val is None:
                    return ""
                if isinstance(val, (dict, list)):
                    return json.dumps(val)
                return str(val)
            return match.group(0)

        return re.sub(r"(?<!\\)\{\{([^#/][^}]*?)\}\}", replace_var, template)

    @staticmethod
    def _is_truthy(value: Any) -> bool:
        if value is None or value is False or value == "" or value == 0:
            return False
        if isinstance(value, list) and len(value) == 0:
            return False
        return True
